docs/rule builder.md

```diff
@@ -0,0 +1,456 @@
+
+The rule builder provides a `CachedRuleBuilderWorld` base class for your `World` class that enables caching on your rules.
+
+```python
+class MyWorld(CachedRuleBuilderWorld):
+    game = "My Game"
+```
+
+If your world's logic is very simple and you don't have many nested rules, the caching system may have more overhead cost than time it saves. You'll have to benchmark your own world to see if it should be disabled or not.
+
+If you enabled caching and your rules use `CanReachLocation`, `CanReachEntrance` or a custom rule that depends on locations or entrances, you must call `self.register_dependencies()` after all of your locations and entrances exist to setup the caching system.
+
+## Item name mapping
+
+If you have multiple real items that map to a single logic item, add a `item_mapping` class dict to your world that maps actual item names to real item names so the cache system knows what to invalidate.
+
+For example, if you have multiple `Currency x<num>` items on locations, but your rules only check a singular logical `Currency` item, eg `Has("Currency", 1000)`, you'll want to map each numerical currency item to the single logical `Currency`.
+
+```python
+class MyWorld(CachedRuleBuilderWorld):
+    item_mapping = {
+        "Currency x10": "Currency",
+        "Currency x50": "Currency",
+        "Currency x100": "Currency",
+        "Currency x500": "Currency",
+    }
+```
+
+## Defining custom rules
+
+You can create a custom rule by creating a class that inherits from `Rule` or any of the default rules. You must provide the game name as an argument to the class. It's recommended to use the `@dataclass` decorator to reduce boilerplate to provide your world as a type argument to add correct type checking to the `_instantiate` method.
+
+You must provide or inherit a `Resolved` child class that defines an `_evaluate` method. This class will automatically be converted into a frozen `dataclass`. You may need to also define one or more dependencies functions as outlined below.
+
+To add a rule that checks if the user has enough mcguffins to goal, with a randomized requirement:
+
+```python
+@dataclasses.dataclass()
+class CanGoal(Rule["MyWorld"], game="My Game"):
+    def _instantiate(self, world: "MyWorld") -> Rule.Resolved:
+        return self.Resolved(world.required_mcguffins, player=world.player, caching_enabled=world.rule_caching_enabled)
+
+    class Resolved(Rule.Resolved):
+        goal: int
+
+        def _evaluate(self, state: CollectionState) -> bool:
+            return state.has("McGuffin", self.player, count=self.goal)
+
+        def item_dependencies(self) -> dict[str, set[int]]:
+            return {"McGuffin": {id(self)}}
+```
+
+### Item dependencies
+
+If there are items that when collected will affect the result of your rule evaluation, it must define an `item_dependencies` function that returns a mapping of the item name to the id of your rule. These dependencies will be combined to inform the caching system.
```

Comment:
This is only true now if you use a CachedRuleBuilderWorld, right? Just testing my own understanding more than anything, I think it's fine to kind of "lie by omission" here

Reply:
Yeah, I never revisited these sections after the split. I can see other systems using these in the future, though (eg fill)

Reply:
Updated these to be more clear about when they're important